### ISSUE TITLE
Concurrency bug

### ISSUE THREAD

**Issue by ansoni-san** [lotusirous/go-concurrency-patterns#11]
https://github.com/lotusirous/go-concurrency-patterns/blob/82c23134af1a6f4a847338e0eadfde940a35a608/17-ring-buffer-channel/main.go#L28

This is a bug surely? Another routine could write between these two calls, and then this write would block.

A safer approach would simply be to drop the current write AND one from the ring buffer. This would ensure that no code ever blocks, but you are still guaranteed new data (which IMO is a big use-case for ring buffers).

Ensuring no back-pressure, as long as your ring buffer isn't too small.

**Comment by attila-kun** [comment]
Agreed, in its current form the code is prone to deadlocking. Proof: https://github.com/lotusirous/go-concurrency-patterns/compare/main...attila-kun:deadlock?expand=1#diff-feadaed927370e3ae128f41c317b1b3f57d3b0b2900612ce96314c4a9fcb9bd8R30

It required the introduction of an artificial delay before `r.outCh <- v ` and a concurrent consumer of `outCh` but if you run the above, the result is a deadlock (might need to run multiple times, it doesn't always happen).
